### ISSUE TITLE
[sklearn#34032] BUG: Silent float32 upcasting to float64 in Pipeline/StandardScaler causes 2x memory bloat

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#361]
**Mirrored from [scikit-learn#34032](https://github.com/scikit-learn/scikit-learn/issues/34032)**

Original title: BUG: Silent float32 upcasting to float64 in Pipeline/StandardScaler causes 2x memory bloat
Original labels: Bug, Needs Investigation
Created: 2026-05-17T01:48:41Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

### Describe the bug and give evidence about its user-facing impact

When processing massive financial tick data (e.g., shape `(10_000_000, 50)`), quant systems typically cast down to `np.float32` to keep the data within RAM limits. 

However, passing a `float32` NumPy array through a `Pipeline` containing a `StandardScaler` silently upcasts the transformed output back to `float64`. This creates a silent memory copy that doubles the RAM requirement, frequently triggering Out-Of-Memory (OOM) kills on high-frequency streaming applications.

### Steps/Code to Reproduce

```python
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

# Generate a mock financial tick dataset in float32
X_financial_ticks = np.random.randn(1_000_000, 10).astype(np.float32)

print(f"Input dtype: {X_financial_ticks.dtype}")
print(f"Input memory: {X_financial_ticks.nbytes / 1e6} MB")

# Pass through a standard pipeline
pipeline = Pipeline([
    ('scaler', StandardScaler())
])

X_transformed = pipeline.fit_transform(X_financial_ticks)

print(f"Output dtype: {X_transformed.dtype}")
print(f"Output memory: {X_transformed.nbytes / 1e6} MB")
```

### Expected Results

**Expected Results:**

```
Input dtype: float32
Input memory: 40.0 MB
Output dtype: float32
Output memory: 40.0 MB
```

### Actual Results

**Actual Results:**
```
Input dtype: float32
Input memory: 40.0 MB
Output dtype: float64
Output memory: 80.0 MB  <-- 100% memory bloat
```

### Versions

```shell
System:
    python: 3.13.11 (main, Dec  8 2025, 11:43:54) [GCC 15.2.0]
executable: /usr/bin/python
   machine: Linux-6.17.10+kali-amd64-x86_64-with-glibc2.42

Python dependencies:
      sklearn: 1.8.0
          pip: 25.3
   setuptools: 78.1.1
        numpy: 2.3.5
        scipy: 1.16.3
       Cython: None
       pandas: 2.3.3
   matplotlib: 3.10.7+dfsg1
       joblib: 1.5.2
threadpoolctl: 3.6.0

Built with OpenMP: True

threadpoolctl info:
       user_api: openmp
   internal_api: openmp
    num_threads: 4
         prefix: libgomp
       filepath: /home/killer/.local/lib/python3.13/site-packages/scikit_learn.libs/libgomp-e985bcbb.so.1.0.0
        version: None
```

### Interest in fixing the bug

**Proposed Solution:**
I would like to investigate the `check_array` / `dtype` handling inside `StandardScaler.transform` to ensure it respects the input dtype without silently upcasting, provided the user explicitly inputs float32.
If the maintainers agree this is an anti-pattern, please assign this to me. I am h

...(truncated)

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [52](https://github.com/godofecht/flow-scikit/issues/52) in the flow-scikit repository. Further library-specific triage and implementation will happen there.

**Comment by godofecht** [comment]
Moved to [53](https://github.com/godofecht/flow-scikit/issues/53) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
